### PR TITLE
Ensure v0.13.0 not installed with nx 2.0

### PR DESCRIPTION
**Pull request by jni**
skimage 0.13.0 is incompatible with networkx 2.0. (0.13.1, just released, coming in separate PR) This PR makes it explicit. But, thinking back on the previous PR, this wasn't a problem with tests... Is this change implied in conda-forge?

### CODE REVIEW

**Comment by Korijn** [comment]
Maybe implicitly through one of the other dependencies?

**Comment by Korijn** [comment]
Also, it doesn't look like we actually run the tests. You can [extend the recipe to run the test suite](https://conda.io/docs/user-guide/tasks/build-packages/define-metadata.html#test-section)?

**Comment by jni** [comment]
@Korijn I think that should do it, but it depends where the tests are run from. Is it from the project root directory? Or somewhere else? `pytest skimage` should point to the `skimage` directory of the source code.

**Comment by jni** [comment]
Also, the library should be built in-place.

**Comment by Korijn** [comment]
Conda switches to an empty (out-of-source) directory before starting the test phase.

I think the tests are included in the installed package? Then you can run with `pytest --pyargs mypkg` ([docs](https://docs.pytest.org/en/latest/goodpractices.html#tests-as-part-of-application-code))

**Comment by jni** [comment]
@Korijn good point! I forgot about that idiom. Hoping this build works!

**Comment by Korijn** [comment]
if getting the test suite to pass here proves to become a large effort you might want to merge your dependency changes first and open another PR for the test suite changes!

File: recipe/meta.yaml [hunk 1]
```diff
@@ -34,19 +34,22 @@ requirements:
...
+    - networkx >=1.8,<2.0
     - pillow >=2.1.0
     - dask-core >=0.15
     - toolz >=0.7.4
     - pywavelets >=0.4.0
     - imageio >=2.1.0
 
 test:
+  requires:
+    - pytest >=3,<4
```
[the remark this answers is not in the document]
Comment (jni) [reply recipe/meta.yaml]: Ah! Sorry, we moved to pytest recently! Apparently *after* 0.13. Apologies, will change!

File: recipe/meta.yaml [hunk 2]
```diff
@@ -34,19 +34,22 @@ requirements:
...
 
 test:
+  requires:
+    - nose
   imports:
     - skimage
   commands:
     - conda inspect linkages -p $PREFIX scikit-image  # [not win]
     - conda inspect objects -p $PREFIX scikit-image  # [osx]
+    - nosetests --stop --no-capture --with-doctest skimage
```
Comment (jakirkham) [inline recipe/meta.yaml]: `--no-capture` -> `--nocapture` (sans `-`)

**Comment by Korijn** [comment]
So should the pillow<4.2 testing requirement also be imposed on the runtime dependencies?

**Comment by jni** [comment]
@Korijn no, it's an incorrectly-written test that causes later versions to fail. This test has been removed from master. But users shouldn't have any issue (except when they use incorrect data, in which case the error is the correct thing to do).

On the other hand, this is a bit worse:

```
File "/feedstock_root/build_artefacts/scikit-image_1506931307892/_t_env/lib/python3.6/site-packages/skimage/run-hessian.py", line 11, in <module>
```

`run-hessian.py` is an old, random test script that the 0.13 release manager (coughs, looks around nervously) inadvertently pushed to PyPI! I honestly don't know what to do with this right now. LOL

**Comment by jakirkham** [comment]
Could just patch out the bad script and broken test.

**Comment by jni** [comment]
I was gonna ask where to do that but it became instantly obvious. (I hope!)

**Comment by Korijn** [comment]
Unfortunately setting environment variables like that won't work on Windows/AppVeyor. :)

```
(C:\bld\scikit-image_1506938164320\_t_env) (root) C:\bld\scikit-image_1506938164320\test_tmp>QT_QPA_PLATFORM=offscreen MPLBACKEND=agg nosetests --stop --nocapture --with-doctest skimage 
'QT_QPA_PLATFORM' is not recognized as an internal or external command,
operable program or batch file.
```

I suppose you'd need to write something like:

```
commands:
  - setx QT_QPA_PLATFORM offscreen  # [win]
  - setx MPLBACKEND agg  # [win]
  - QT_QPA_PLATFORM=offscreen  # [not win]
  - MPLBACKEND=agg  # [not win]
  - nosetests --stop --nocapture --with-doctest skimage
```

**Comment by jni** [comment]
Please excuse me while I rediscover everything going on in [this script](https://github.com/scikit-image/scikit-image/blob/master/tools/travis/script.sh)...

**Comment by jni** [comment]
@Korijn thanks, noted! Looks like it'll need a few more lines than that, too. Thanks for your patience! I'm starting to see why this recipe didn't run the tests in the first place... =P

**Comment by Korijn** [comment]
@jni community service :+1:

**Comment by jni** [comment]
Hmm, here's the current traceback, apparently unrelated to skimage... Is there something wrong with the recipe now?

```pytb
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/opt/conda/lib/python3.6/site-packages/conda_build/cli/main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.6/site-packages/conda_build/cli/main_build.py", line 333, in execute
    noverify=args.no_verify)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1530, in build_tree
    test(pkg, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1299, in test
    shell_files = create_shell_files(config.test_dir, metadata, config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/create_test.py", line 80, in create_shell_files
    f.write(cmd)
TypeError: write() argument must be str, not dict
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -34,19 +35,30 @@ requirements:
...
   commands:
     - conda inspect linkages -p $PREFIX scikit-image  # [not win]
     - conda inspect objects -p $PREFIX scikit-image  # [osx]
+    - set MPLCONFIGDIR .  # [win]
+    - set QT_QPA_PLATFORM offscreen  # [win]
+    - set MPLBACKEND Qt4agg  # [win]
+    - MPL_CONFIG_DIR=.  # [not win]
+    - QT_QPA_PLATFORM=offscreen  # [not win]
+    - MPLBACKEND=Qt4agg  # [not win]
+    - echo 'backend.qt : PyQt4' > matplotlibrc
```
Comment (jakirkham) [inline recipe/meta.yaml]: Would put this line in quotes (everything after the `-`). I think the `:` is getting treated as YAML hence the error message.

**Comment by jakirkham** [comment]
It may be easier to move the testing code into `run_test.bat` and `run_test.sh`. Alternatively could try `run_test.py`. This should help keep the `meta.yaml` cleaner, make it easier when dealing with these platform specific things, and avoid things being parsed as YAML when they shouldn't be.

**Comment by Korijn** [comment]
To be clear: you don't need to call those scripts from the recipe. Just the presence of those files should be enough, right @jakirkham ?

**Comment by jakirkham** [comment]
Correct. `conda-build` copies them into the testing directory and calls them for you.

File: recipe/meta.yaml [hunk 4]
```diff
@@ -9,9 +9,10 @@ source:
   sha256: 77a636bdc08c7668a15951894548c527f0c8c5c2abc86cb850de17551af51e3e
 
 build:
-  number: 2
+  number: 3
   script:
     - export LDFLAGS="-headerpad_max_install_names"  # [osx]
```
Comment (jakirkham) [inline recipe/meta.yaml]: We should be able to drop this line if we add `toolchain` to `build`. I should have done that a long time ago.

**Comment by jakirkham** [comment]
> I'm starting to see why this recipe didn't run the tests in the first place... =P

Yeah that's my bad. I think I tried locally when initially writing this recipe and had a rough time of it. Should have asked for help sooner. Sorry.

**Comment by jni** [comment]
@jakirkham didn't know whether to place `run_tests.py` in the `recipe` directory, let me know if that needs to change. Also, if anyone has suggestions for how to iterate faster (instead of pushing and waiting for CircleCI to run things, spamming everyone in the process), I'm all ears!

**Comment by jakirkham** [comment]
Yeah it needs to go in the `recipe` directory. When it comes to the feedstocks, everything outside the `recipe` directory is autogenerated. So things put there tend to get overwritten.

Can use the [`ci_support/run_docker_build.sh`]( https://github.com/conda-forge/scikit-image-feedstock/blob/master/ci_support/run_docker_build.sh ). That would let you run the Linux build locally in a Docker container. Will need to install Docker if you don't have it, but should be straightforward. Also need to supply the `CONDA_PY` environment variable (e.g. `export CONDA_PY=36` for a Python 3.6 build).

**Comment by jni** [comment]
Woot! =) Not sure why AppVeyor is failing, only one failure in the whole build matrix...

File: recipe/meta.yaml [hunk 5]
```diff
@@ -34,14 +36,17 @@ requirements:
...
+    - networkx >=1.8,<2.0
     - pillow >=2.1.0
     - dask-core >=0.15
     - toolz >=0.7.4
     - pywavelets >=0.4.0
     - imageio >=2.1.0
 
 test:
+  requires:
+    - pillow <4.2
```
Comment (jakirkham) [inline recipe/meta.yaml]: Should there be a lower bound too?

Comment (jni) [reply recipe/meta.yaml]: Does the lower bound not propagate from the runtime requirements...?

File: recipe/meta.yaml (hunk 4 above)
Comment (jni) [reply recipe/meta.yaml]: I don't know what this comment means but it seems to be very suitable for a separate PR... =P

File: recipe/meta.yaml (hunk 5 above)
Comment (jakirkham) [reply recipe/meta.yaml]: Good point. Overlooked that. Should be fine.

File: recipe/meta.yaml (hunk 4 above)
Comment (jakirkham) [reply recipe/meta.yaml]: Yeah I can address it in a follow-up PR.

**Comment by Korijn** [comment]
I think you named the file wrong; should be `run_test.py`, not `run_tests.py`. It was skipped now.

**Comment by jni** [comment]
@Korijn I thought it was amazing that the tests passed first try!

**Comment by Korijn** [comment]
It made me suspicious! :D

**Comment by Korijn** [comment]
run_test.py does not return the exit code of the test suite it seems, because it failed even though the build status is green. :(

**Comment by jni** [comment]
@Korijn @jakirkham things are looking good I think! =)

**Comment by jakirkham** [comment]
Is there some verbosity setting for the test suite? Seems it is so quiet I can't tell if it actually ran. 😕

**Comment by jni** [comment]
@jakirkham the line "Ran 1465 tests in 56.095s" is very suggestive, but turned it on anyway. =)

**Comment by Korijn** [comment]
@jni things failed again! :(

**Comment by jni** [comment]
@Korijn they are random failures. I've restarted Travis but don't seem to have that power with AppVeyor. But if you look at av all passed except one build.

**Comment by Korijn** [comment]
@jni I've gone ahead and restarted the build for you.

When signing in at AppVeyor, do so through your Github credentials. You will be prompted to pick an organization to sign in with, and conda-forge will be one of them. That will give you the permissions to restart builds for conda-forge.

**Comment by jni** [comment]
@Korijn thanks for the tip! Restarted the build yet again. Reproducible error on 3.6/x64, but appears unrelated to scikit-image... =\

https://ci.appveyor.com/project/conda-forge/scikit-image-feedstock/build/1.0.65/job/lo7hia8ogpd52n6y#L2896
```pytb
subprocess.CalledProcessError: Command '['cmd.exe', '/c', 'bld.bat']' returned non-zero exit status 1.
```
<details>
<pre>
Command "C:\bld\scikit-image_1508224900357\_b_env\python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\pip-mwrf1hb8-build\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\appveyor\AppData\Local\Temp\1\pip-mv_qkvl4-record\install-record.txt --single-version-externally-managed --compile" failed with error code 3221225477 in C:\Users\appveyor\AppData\Local\Temp\1\pip-mwrf1hb8-build\
Traceback (most recent call last):
  File "C:\Miniconda36-x64\Scripts\conda-build-script.py", line 5, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\cli\main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\cli\main_build.py", line 333, in execute
    noverify=args.no_verify)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\build.py", line 1524, in build_tree
    config=config)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\build.py", line 1114, in build
    windows.build(m, build_file, config=config)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\windows.py", line 231, in build
    check_call_env(cmd, cwd=src_dir)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\utils.py", line 628, in check_call_env
    return _func_defaulting_env_to_os_environ(subprocess.check_call, *popenargs, **kwargs)
  File "C:\Miniconda36-x64\lib\site-packages\conda_build\utils.py", line 624, in _func_defaulting_env_to_os_environ
    return func(_args, **kwargs)
  File "C:\Miniconda36-x64\lib\subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['cmd.exe', '/c', 'bld.bat']' returned non-zero exit status 1.
</pre>
</details>

**Comment by jni** [comment]
Oh, not sure yet if it's reproducible, it's queued, needed a browser refresh... Fingers crossed! 🤞

**Comment by jni** [comment]
Ok so this time both 35 and 36 failed, same error. I'm going to assume it's some high probability network error? Running again, but if anyone feels like just merging, I won't stop them... =P

**Comment by jakirkham** [comment]
> the line "Ran 1465 tests in 56.095s" is very suggestive, but turned it on anyway. =)

Must have missed it before. Definitely see it now. Thanks. :)

**Comment by jakirkham** [comment]
This seems relevant.

```batch
python.exe -u -c "import setuptools, tokenize;__file__='C:\\Users\\appveyor\\AppData\\Local\\Temp\\1\\pip-dns5mvk1-build\\setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record C:\Users\appveyor\AppData\Local\Temp\1\pip-6dlc7uq0-record\install-record.txt --single-version-externally-managed --compile"
```

ref: https://ci.appveyor.com/project/conda-forge/scikit-image-feedstock/build/1.0.67/job/r60lbrl1929obsyb#L2874

File: recipe/meta.yaml [hunk 6]
```diff
@@ -9,9 +9,11 @@ source:
   sha256: 77a636bdc08c7668a15951894548c527f0c8c5c2abc86cb850de17551af51e3e
 
 build:
-  number: 2
+  number: 3
   script:
     - export LDFLAGS="-headerpad_max_install_names"  # [osx]
+    - rm -f skimage/run-hessian.py
+    - rm -rf skimage/viewer/tests  # we don't depend on Qt
     - python -m pip install --no-deps .
```
Comment (jakirkham) [inline recipe/meta.yaml]: What if we trying going back to `python setup.py install --single-version-externally-managed --record record.txt`?

Comment (jni) [reply recipe/meta.yaml]: @jakirkham can you explain why you think this is failing?

**Comment by jni** [comment]
@jakirkham I've changed as you suggested but I don't understand why... Any clarification would be appreciated.

**Comment by jakirkham** [comment]
Honestly it was just a gut feeling after looking at the error that this would be something easy to try that might work. Happy to see that it did.

**Comment by jni** [comment]
@jakirkham @Korijn merge time???

**Comment by jakirkham** [comment]
Yep. Wasn't sure if we were going to discuss the change to the build script further or not. Happy to merge as is.

**Comment by jakirkham** [comment]
Thanks @jni! Also thanks @Korijn for helping review.

**Comment by jakirkham** [comment]
Odd 😕 

```
======================================================================
ERROR: skimage.morphology.tests.test_skeletonize.TestThin.test_noiter
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\bld\scikit-image_1509516280578\_t_env\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\bld\scikit-image_1509516280578\_t_env\lib\site-packages\skimage\morphology\tests\test_skeletonize.py", line 144, in test_noiter
    result = thin(self.input_image).astype(np.uint8)
  File "C:\bld\scikit-image_1509516280578\_t_env\lib\site-packages\skimage\morphology\_skeletonize.py", line 266, in thin
    for i in range(max_iter):
OverflowError: Python int too large to convert to C long
======================================================================
ERROR: skimage.morphology.tests.test_skeletonize.TestThin.test_zeros
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\bld\scikit-image_1509516280578\_t_env\lib\site-packages\nose\case.py", line 197, in runTest
    self.test(*self.arg)
  File "C:\bld\scikit-image_1509516280578\_t_env\lib\site-packages\skimage\morphology\tests\test_skeletonize.py", line 130, in test_zeros
    assert np.all(thin(np.zeros((10, 10))) == False)
  File "C:\bld\scikit-image_1509516280578\_t_env\lib\site-packages\skimage\morphology\_skeletonize.py", line 266, in thin
    for i in range(max_iter):
OverflowError: Python int too large to convert to C long
----------------------------------------------------------------------
```

ref: https://ci.appveyor.com/project/conda-forge/scikit-image-feedstock/build/1.0.70/job/a4qafr8vvhc2x3pt#L5229

**Comment by jni** [comment]
https://github.com/scikit-image/scikit-image/pull/2846

haven't investigated the source of this problem yet...

**Pull request merged**